### PR TITLE
Allow markdown files to read from the LFS

### LINKED ISSUE #5746: Git LFS image linking not possible
- Gitea version (or commit ref): 1.7 (Downloaded from https://dl.gitea.io/gitea/1.7/) (8006b1b in footer of the instance)
- Git version: 2.17.1
- Operating system: Ubuntu 18.04
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No
  - [ ] Not relevant
- Log gist: -

## Description
When linking to an image that is stored in LFS, you get the LFS text. For example:
https://mywebsite.com/diondokter/myrepo/raw/branch/develop/Design/Images/myimage.jpg gives
> version https://git-lfs.github.com/spec/v1
> oid sha256:e31e396000df86f9a888eea93efec5ad46fa68c1d3f8d6aa8f42991e137deaad
> size 14365

When linking to the image in markdown, it will not appear which is quite inconvenient.
Can this be resolved?

I can't repro this issue on try.gitea.io because (I suspect) LFS is turned off.

> image

### PR DESCRIPTION
**Pull request by zeripath**
This PR makes it possible for the markdown renderer to render images and media straight from the LFS.

Fix #5746 

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/lfs/utils.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	if isTextFile && setting.LFS.StartServer {
+		return IsLFSPointerFile(&buf), &buf
+	}
+
+	return nil, &buf
+}
+
+// IsLFSPointerFile will return a partially filled LFSMetaObject if the provided byte slice is a pointer file
+func IsLFSPointerFile(buf *[]byte) *models.LFSMetaObject {
+	if setting.LFS.StartServer {
```
Comment (lafriks) [inline modules/lfs/utils.go]: Please change this function to make it better readable with reverse conditions, ex:
```go
if !setting.LFS.StartServer {
     return nil
}
headString := string(*buf)
if !strings.HasPrefix(headString, models.LFSMetaFileIdentifier) {
    return nil
}
// etc
```

**Review by lafriks — APPROVED** [review]

**Comment by zeripath** [comment]
~~As there are no tests of this functionality at present I've marked this WIP until I've put some up~~ ~~one basic test implemented~~ Have wired in tests in to TestGit. I think that should be enough.

File: routers/repo/download.go [hunk 2]
```diff
@@ -55,6 +58,23 @@ func ServeBlob(ctx *context.Context, blob *git.Blob) error {
...
 
+// ServeBlobOrLFS download a git.Blob redirecting to LFS if necessary
+func ServeBlobOrLFS(ctx *context.Context, blob *git.Blob) error {
+	dataRc, err := blob.DataAsync()
+	if err != nil {
+		return err
+	}
+	defer dataRc.Close()
+
+	if meta, _ := lfs.ReadLFSPointerFile(dataRc); meta != nil {
```
[the remark this answers is not in the document]
Comment (zeripath) [reply routers/repo/download.go]: The rights are checked earlier in the function chain and the oid+repo right table would have been checked in the redirect - but I've just removed that redirect so am checking that now.

**Comment by zeripath** [comment]
~~OK it appears that creating tests for this might be a bit more involved so I'm going to remove the WIP.~~

~~If proper integration tests are required please ask and I'll add some more.~~

More tests wired into TestGit .

**Comment by lunny** [comment]
@zeripath please add integration test

**Comment by zeripath** [comment]
@techknowlogick I think that's the only sensible way to do it - I'm just loathe to make PR dependencies. 

Our integration tests of the editor view are also somewhat lacking, as are those of the `/raw` endpoints and so on. This basically means that in order to get this 200 line change in, I need to write tests for at least 2000 lines of code.

I guess that's technical debt for you.

**Comment by zeripath** [comment]
Of course if #5722 was merged I'd actually have the beginnings of an LFS test framework to work with...

**Comment by zeripath** [comment]
OK the /raw and /media pathways are now checked in TestGit.

**Review by lafriks — APPROVED** [review]

File: modules/lfs/utils.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
```
Comment (lunny) [inline modules/lfs/utils.go]: The file name `utils.go` is not a good name.

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/lfs/utils.go (hunk 3 above)
Comment (zeripath) [reply modules/lfs/utils.go]: I've changed it to lfs_pointers.go

**Comment by zeripath** [comment]
I've removed the `/wiki/media` pathways as at present wiki files can't be placed in to the LFS and #5814 shows that a simple `/wiki/media` url route would need to reserve `media` from the wiki.

**Comment by zeripath** [comment]
Ping @lunny

**Comment by lunny** [comment]
@zeripath I think the file names could be more simpler as `pointer.go` and the methods could remove `LFS` since all these are on package `lfs`. When you use them, you will reference them as `lfs.ReadLFSPointerFile` so the duplicated `lfs` on  method name is unnecessary.

**Comment by zeripath** [comment]
OK, I've done that.

**Comment by zeripath** [comment]
@lunny any more work needed on this?

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
This PR makes it possible for the markdown renderer to render images and media straight from the LFS.

Fix #5746

Signed-off-by: Andrew Thornton [art27@cantab.net](mailto:art27@cantab.net)

**Pull request merged**